src/compiler/checker.ts

```diff
@@ -10343,6 +10343,10 @@ namespace ts {
             if (!type) {
                 return undefined;
             }
+            // If the contextual signature has fewer parameters than the function expression, do not use it
+            if (isFunctionExpressionOrArrowFunction(node) && isAritySmaller(type, node)) {
+                return undefined;
+            }

```

Comment:
I moved the check inside getNonGenericSignature so that it gets called in both the union and non-union paths. Let me know if that looks good to you.